### PR TITLE
Create Lessons Endpoints

### PR DESCRIPTION
**Pull request by nickerick**
## Accomplishments
- Create Lessons model and initial endpoints
- Various cleanup
- Added jakarta validation
- Added global exception handlers

### Create Course null fields check

Doing `curl -X POST localhost:8080/api/course -d {}`
Returns a 400 "Course does not have all required fields"

---

### Create Lesson missing fields check

Doing `curl -X POST localhost:8080/api/lessons -d { "title" : null }`
Returns 400
```json
{
    "videoType": "Must supply either SELF_HOSTED or YOUTUBE",
    "videoNumber": "Must supply a valid lesson number",
    "videoLink": "must not be empty",
    "courseId": "Must supply a valid course id"
}
```

### Create Lesson
Doing
```curl
curl -X POST localhost:8080/api/lessons -d { "courseId":1, "title":"Overview 1", "videoLink":"myUrl.com", "videoType":"SELF_HOSTED", "videoNumber":1 }
```
Returns 200
```json
{
    "id": 3,
    "title": "Overview 2",
    "videoLink": "myUrl.com",
    "videoType": "SELF_HOSTED",
    "videoNumber": 1
}
```
### Delete Lesson
Doing`curl -X DELETE localhost:8080/api/lessons/1`
Returns 200
```json
{
    "id": 3,
    "title": "Overview 2",
    "videoLink": "myUrl.com",
    "videoType": "SELF_HOSTED",
    "videoNumber": 1
}
```
### Update Lesson
Doing`curl -X PUT localhost:8080/api/lessons/1` -d { "title" : "NEW TITLE", "videoLink" : "newlink.com" }
Returns 200
```json
{
    "id": 3,
    "title": "NEW TITLE",
    "videoLink": "newlink.com",
    "videoType": "YOUTUBE",
    "videoNumber": 1
}
```

### Get Courses
Doing `curl -X GTE localhost:8080/api/courses`
Returns 200
```json
[
    {
        "id": 2,
        "lessons": [
            {
                "id": 3,
                "title": "Overview 2",
                "videoLink": "myUrl.com",
                "videoType": "SELF_HOSTED",
                "videoNumber": 1
            }
        ],
        "title": "Test Course 2",
        "description": "super cool test course",
        "category": "programming",
        "views": 0,
        "thumbsUp": 0,
        "thumbsDown": 0,
        "createdAt": "2024-02-22T01:20:54.715175",
        "status": "2024-02-22T01:20:54.715175"
    },
    {
        "id": 52,
        "lessons": [],
        "title": "Test Course 3",
        "description": "super cool test course",
        "category": "programminasdfsadfag",
        "views": 0,
        "thumbsUp": 0,
        "thumbsDown": 0,
        "createdAt": "2024-02-22T01:29:34.195824",
        "status": "2024-02-22T01:29:34.195824"
    }
]
```

### CODE REVIEW

File: server/src/main/java/com/skillmagnet/Course/Course.java [hunk 1]
```diff
@@ -1,47 +1,43 @@
...
-import jakarta.persistence.GeneratedValue;
-import jakarta.persistence.Id;
-import jakarta.persistence.Table;
-import lombok.Data;
+import java.time.LocalDateTime;
+import java.util.List;
 
-@Data
+@Getter
+@Setter
```
Comment (nickerick) [inline server/src/main/java/com/skillmagnet/Course/Course.java:13]: :memo: Did some cleanup on the lombok stuff throughout

File: server/src/main/java/com/skillmagnet/Course/CourseController.java [hunk 2]
```diff
@@ -20,75 +20,82 @@ public class CourseController {
 
     /**
      * Retrieves specific course by id
+     *
```
Comment (nickerick) [inline server/src/main/java/com/skillmagnet/Course/CourseController.java:23]: :memo: there's lots of miscellaneous spaces and formatting that got applied from my autoformatter

File: server/src/main/java/com/skillmagnet/Course/CourseController.java [hunk 3]
```diff
@@ -20,75 +20,82 @@ public class CourseController {
...
     @PostMapping("/course")
-    public ResponseEntity<Course> createCourse(@RequestBody Course course){
+    public ResponseEntity<?> createCourse(@RequestBody Course course) {
+        if (null == course.getTitle() || null == course.getCategory() || null == course.getDescription()) {
+            return new ResponseEntity<>("Course does not have all required fields", HttpStatus.BAD_REQUEST);
+        }
+
         courseRepository.save(course);
         return new ResponseEntity<>(course, HttpStatus.OK);
     }
```
Comment (nickerick) [inline server/src/main/java/com/skillmagnet/Course/CourseController.java:69]: :memo: Added check to make sure title, category, and description were not null. The other properties autogenerate or default to 0

File: server/src/main/java/com/skillmagnet/GlobalExceptionHandler.java
Comment (nickerick) [inline server/src/main/java/com/skillmagnet/GlobalExceptionHandler.java:1]: :memo: Global exception handlers. So basically if one of these exceptions is thrown and the endpoint doesn't directly catch it in a try/catch, it'll re-route to here.

MethodArgumentNotValidException is thrown when Spring's validation fails on properties

HttpMessageNotReadableException is thrown when failure on serializing to JSON to Objects. Added in to handle Enum mapping

Comment (sjdefran1) [reply server/src/main/java/com/skillmagnet/GlobalExceptionHandler.java:1]: woahhh this is sick, me llikey

File: server/src/main/java/com/skillmagnet/Lesson/Lesson.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.skillmagnet.Course.Course;
+import jakarta.persistence.*;
+import lombok.*;
+
+@Getter
+@Setter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
Comment (sjdefran1) [inline server/src/main/java/com/skillmagnet/Lesson/Lesson.java:12]: I saw you changed from `@Data` to `@Getter & @Setter` any specific reason? No problems with it, just wondering. I know `@Data` includes these annotations, read up on it [here](https://projectlombok.org/features/Data)

File: server/src/main/java/com/skillmagnet/Lesson/Lesson.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+@Table(name = "lesson")
+public class Lesson {
+
+    @Id
+    @GeneratedValue
+    private int id;
+
+    @ManyToOne
+    @JoinColumn(name = "course_id")
+    private Course course;
```
Comment (sjdefran1) [inline server/src/main/java/com/skillmagnet/Lesson/Lesson.java:23]: What does `name = "course_id"` refer to. Because the SQL table is just id correct? as well as in the class for Course where it also id. Does it know since its a `Course` object? Is this some sort of syntax magic?

File: server/src/main/java/com/skillmagnet/Lesson/LessonController.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+    /**
+     * Retrieves a lesson by its ID.
+     *
+     * @param id The ID of the lesson to retrieve.
+     * @return {@code Lesson} on success otherwise a NOT_FOUND status.
+     */
+    @GetMapping("/lessons/{id}")
+    public ResponseEntity<?> getLesson(@PathVariable("id") int id) {
+        Optional<Lesson> lesson = lessonRepository.findById(id);
```
Comment (sjdefran1) [inline server/src/main/java/com/skillmagnet/Lesson/LessonController.java:32]: Should I be using Optional<> ? What is the advantage? We can still check for null if it was just `Lesson lesson`? better standard?

File: server/src/main/java/com/skillmagnet/Lesson/LessonRequest.java
Comment (sjdefran1) [inline server/src/main/java/com/skillmagnet/Lesson/LessonRequest.java:1]: Ok  this is cool, we should add these to other existing Entities. Definitely eliminates old constructor confusion

File: server/src/main/java/com/skillmagnet/Lesson/LessonRepository.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package com.skillmagnet.Lesson;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface LessonRepository extends JpaRepository<Lesson, Integer> {
```
Comment (sjdefran1) [inline server/src/main/java/com/skillmagnet/Lesson/LessonRepository.java:7]: I was going to ask about this, using `<Lesson, Integer>` rather then `<Lesson, Long>`. I like it, avoids having to create `findById(int i)`

File: server/src/main/java/com/skillmagnet/Course/Course.java [hunk 8]
```diff
@@ -1,47 +1,43 @@
...
 @Entity
 @Table(name = "course")
 public class Course {
+
     @Id
     @GeneratedValue
     private int id;
 
+    @OneToMany(mappedBy = "course", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
+    private List<Lesson> lessons;
```
Comment (sjdefran1) [inline server/src/main/java/com/skillmagnet/Course/Course.java:23]: Is it safe to assume we should always use `cascade = CascadeType.ALL`? implementing something similar for enrolls

**Review by sjdefran1 — COMMENTED** [review]
Looking good! Left some general questions here and there for my own sake. Otherwise I think this looks pretty much good to go

File: server/src/main/java/com/skillmagnet/Lesson/Lesson.java (hunk 4 above)
Comment (nickerick) [reply server/src/main/java/com/skillmagnet/Lesson/Lesson.java:12]: There's no proper reason - more just personal taste choice lol. The `@Data` does include all of that stuff, but also adds toString(), toEquals() and some other stuff that I'm not actually implementing. While functionally it changes absolutely nothing, I just like seeing all the different effects on my class being very verbose and accurate to what I want rather than lumping in bunch of stuff with `@Data`.

TLDR: I prefer extremely verbose code at the tradeoff of taking more lines. Completely debatable

File: server/src/main/java/com/skillmagnet/Lesson/Lesson.java (hunk 5 above)
Comment (nickerick) [reply server/src/main/java/com/skillmagnet/Lesson/Lesson.java:23]: Yeah there's Spring magic on here. I'm not 100% on this, but what I believe is going on is: `@ManyToOne` is the actual mapping magic. That's just gonna automatically map course IDs to lesson objects. The `@JoinColumn` is telling the name of the column on the Lesson object. So in the db the courseId will show up as `course_id` however I could make it something like `course_zebra_cakes` and it'd function exactly the same minus different name in mysql

File: server/src/main/java/com/skillmagnet/Lesson/LessonRequest.java
Comment (nickerick) [reply server/src/main/java/com/skillmagnet/Lesson/LessonRequest.java:1]: Yeah the user might be fine but was considering doing the Course object in this PR but decided was out of scope. Also cleans up the swagger docs

File: server/src/main/java/com/skillmagnet/Lesson/LessonController.java (hunk 6 above)
Comment (nickerick) [reply server/src/main/java/com/skillmagnet/Lesson/LessonController.java:32]: That's a good question and I don't really know. There are definitely cases and like code style guides that would suggest to always use Optionals. It can be super powerful / necessary in more complex spots. 

For how I'm using it here, it's acting no different than just doing `myObject != null` but the `lessonRepository` default `findById()` returns an optional thus why I'm using it lol

**Review by sjdefran1 — APPROVED** [review]
